core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -159,40 +159,18 @@ static List<RelDataType> fieldRowTypes(
 
   static Expression joinSelector(JoinRelType joinType, PhysType physType,
       List<PhysType> inputPhysTypes) {
-    // A parameter for each input.
-    final List<ParameterExpression> parameters = new ArrayList<>();
-
-    // Generate all fields.
-    final List<Expression> expressions = new ArrayList<>();
     final int outputFieldCount = physType.getRowType().getFieldCount();
-
     // If there are many output fields, create the output dynamically so that the code size stays
     // below the limit. See CALCITE-3094.
-    final boolean generateCompactCode = outputFieldCount >= 100;
-    final ParameterExpression compactOutputVar;
-    final BlockBuilder compactCode = new BlockBuilder();
-    if (generateCompactCode) {
-      Class<?> fieldClass = physType.fieldClass(0);
-      // If all fields have the same type, use the specific type. Otherwise just use Object.
-      for (int fieldIndex = 1; fieldIndex < outputFieldCount; ++fieldIndex) {
-        if (fieldClass != physType.fieldClass(fieldIndex)) {
-          fieldClass = Object.class;
-          break;
-        }
-      }
-
-      final Class<?> arrayClass = Array.newInstance(fieldClass, 0).getClass();
-      compactOutputVar = Expressions.variable(arrayClass, "outputArray");
-      final DeclarationStatement exp =
-          Expressions.declare(
-              0, compactOutputVar, new NewArrayExpression(fieldClass, 1,
-              Expressions.constant(outputFieldCount), null));
-      compactCode.add(exp);
-    } else {
-      compactOutputVar = null;
+    if (outputFieldCount >= 100) {
```

Comment:
I think it's a good idea. 

I also don't like this hardcoded there. Is there a better, centralized, place to put this? Then we could also generate the test tables using it, instead of having to hardcode it again in the test class.

Reply:
Maybe we could define a new CalciteSystemProperty of type Integer, perhaps with defaultValue 100, but if the users sets it to 0 then the compactCode version is disable? (just to have a way to "disable" this feature and fallback to the original code in case a problem is found with the compactCodeGenerator afterwards). WDYT?

Reply:
I like the idea. Then it becomes easier also to run all the tests with a specific value, or to benchmark a real use case with different values. If nobody opposes, I can do it.

Reply:
Thanks @rorueda . The CI checks with  the temporary `(outputFieldCount > 0)` threshold have passed successfully, so the new compact code seems stable 👏 
I guess the only remaining bit is reverting this temp change, setting the threshold via a new CalciteSystemProperty of type int, with defaultValue 100 and if <0 then codeCompact is disable. WDYT?

Reply:
I think I will be able to do this at the end of the day. Any suggestion for the property name?

Reply:
I'd propose something along the lines:
```
  /**
   * Minimum numbers of fields in a Join result that will trigger the "compact code generation".
   * This feature reduces the risk of running into a compilation error due to the code of a
   * dynamically generated method growing beyond the 64KB limit.
   *
   * <p>The default value is 100, a negative value disables completely the "compact code" feature.
   *
   * @see org.apache.calcite.adapter.enumerable.EnumUtils
   */
  public static final CalciteSystemProperty<Integer> JOIN_SELECTOR_COMPACT_CODE_THRESHOLD =
      intProperty("calcite.join.selector.compact.code.threshold", 100);
```

Reply:
Also, please verify that setting the new CalciteSystemProperty to e.g. `-1` causes all test to pass, except for LargeGeneratedJoinTest (which should fail since it should reach the 64KB error if compact code is not used)

Reply:
I used your code to add the property, only adding a note on performance. Also had to modify the test to fail when the property is set to `-1`.

To run tests disabling compact code generation:
`./gradlew -Dcalcite.join.selector.compact.code.threshold=-1 test`

To run tests enabling compact code generation for any join size:
`./gradlew -Dcalcite.join.selector.compact.code.threshold=0 test`